libcaf_core/caf/flow/byte.hpp

```diff
@@ -54,6 +56,104 @@ class to_chunks_step {
   byte_buffer buf_;
 };
 
+/// Turns a sequence of bytes into a sequence of chunks based on separator.
+class split_at_step {
+public:
+  using input_type = std::byte;
+
+  using output_type = chunk;
+
+  explicit split_at_step(std::byte separator) noexcept : separator_(separator) {
+    // nop
+  }
+
+  template <class Next, class... Steps>
+  bool on_next(const std::byte& b, Next& next, Steps&... steps) {
+    if (b == separator_)
+      return do_emit(next, steps...);
+    buf_.push_back(b);
+    return true;
+  }
+
+  template <class Next, class... Steps>
+  void on_complete(Next& next, Steps&... steps) {
+    if (!buf_.empty() && !do_emit(next, steps...))
+      return;
+    next.on_complete(steps...);
+  }
+
+  template <class Next, class... Steps>
+  void on_error(const error& what, Next& next, Steps&... steps) {
+    if (!buf_.empty() && !do_emit(next, steps...))
+      return;
+    next.on_error(what, steps...);
+  }
+
+private:
+  template <class Next, class... Steps>
+  bool do_emit(Next& next, Steps&... steps) {
+    auto item = chunk{buf_};
+    buf_.clear();
+    return next.on_next(item, steps...);
+  }
+
+  std::byte separator_;
+  byte_buffer buf_;
+};
+
+/// Turns a sequence of bytes into a sequence of cow_string based on separator.
+class split_as_utf8_at_step {
+public:
+  using input_type = std::byte;
+
+  using output_type = cow_string;
+
+  explicit split_as_utf8_at_step(char separator) noexcept
+    : separator_(separator) {
+    // nop
+  }
+
+  template <class Next, class... Steps>
+  bool on_next(const std::byte& b, Next& next, Steps&... steps) {
+    if (static_cast<char>(b) == separator_)
+      return do_emit(next, steps...);
+    buf_.push_back(b);
+    return true;
+  }
+
+  template <class Next, class... Steps>
+  void on_complete(Next& next, Steps&... steps) {
+    if (!buf_.empty() && !do_emit(next, steps...))
+      return;
+    next.on_complete(steps...);
+  }
+
+  template <class Next, class... Steps>
+  void on_error(const error& what, Next& next, Steps&... steps) {
+    if (!buf_.empty() && !do_emit(next, steps...))
+      return;
+    next.on_error(what, steps...);
+  }
+
+private:
+  template <class Next, class... Steps>
+  bool do_emit(Next& next, Steps&... steps) {
+    if (!detail::rfc3629::valid(buf_)) {
```

Comment:
Should shortcut empty items:

```c++
if (buf_.empty()) {
    return next.on_next(cow_string{} steps...);
}
```

I don't think I saw a test for that case.

Reply:
I have added a new test which checks that empty `cow_string`s are passed to `on_next`.